README
Autobench - an HTTP benchmarking suite

Copyright (C) 2001-2003 Julian T. J. Midgley <jtjm@xenoclast.org>

Homepage: http://www.xenoclast.org/autobench/

Licence
-------

Autobench is distributed under the terms of the GNU Public Licence, a
copy of which may be found in the file "LICENCE".

Manifest
--------

The 'autobench' package comprises the following programs:
  
   autobench
     - A command-line wrapper for httperf which runs httperf a number
       of times against a target host, producing a CSV or TSV format
       file for later analysis/graphing.

   autobench_admin 
     - The administrative controller for distributed autobench.  Used
       to issue commands to a cluster of machines running autobenchd,
       in order to conduct synchronised multiple-client benchmarks
       against a target host.  Collates the results from all the
       clients in the cluster, and outputs CSV or TSV format file as
       with autobench.

   autobenchd
     - The client dameon for distributed autobench.  Listens for
       commands from autobench_admin, and runs httperf to benchmark
       the target host.  Summarises the results and returns them to
       autobench_admin for collation.

   bench2graph
     - Takes TSV benchmark results files produced by autobench or
       autobench_admin and plots them using gnuplot.

   crfile
     - small utility for rapidly creating files filled with random
       data of a specified size.  Useful for creating data sets for
       benchmarking.

   sesslog
     - Creates session logs for use with the --wsesslog argument of
       httperf from NCSA Combined or NCSA Common Log Format web access
       logs.  Useful for generating real-life workloads from your log
       files.

Man pages are included for all of the above.

Installation
------------

See the file "INSTALL" for installation instructions.

Use
---

Autobench can be used in one of two modes.  In the non-distributed
mode, a single autobench client runs benchmarks against one or two
servers.  See the "autobench(1)" man page for details of this mode.

The distributed mode allows a server to be benchmarked using several
clients simultaneously (reducing the risk that the clients themselves
will become the bottleneck).  See the "autobench_admin(1)" and
"autobenchd(1)" man pages for details.


ChangeLog
2012-03-31  Julian T. J. Midgley  <jtjm@menavaur.org>

	* autobench_admin now (for values of "now" >~ 19 Aug 2009) reports
	counts of HTTP status responses received (grouped into hundreds).
	Requested by Ricardo Figueiredo.

2004-05-27  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* Released autobench-2.1.2

	* Added warning about synchronising clocks to autobench_admin.1
	Revision: autobench--trunk--2.1--patch-1
        modified files: autobench_admin.1

	* Autobench now maintained in GNU Arch.  Mirror of archive at:
	http://arch.xenoclast.org/phaedrus-2004/
	Latest version may be fetched with:
	  tla register-archive http://arch.xenoclast.org/phaedrus-2004/
	  tla get autobench--trunk--2.1 autobench

2003-11-25  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* Released [autobench-2-1-1]

	* autobench, autobench.1, autobench_admin:
	Added support for httperf options that don't take an argument (such as
	--ssl).  See the autobench man page for details.

2003-07-27  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* autobench, autobench.1, autobench_admin:
	Added new option --const_test_time which allows all tests to be forced
	to take approximately the same real time; this prevents the early
	tests in a series taking several times longer than the later ones, and
	makes it easier to ensure that the test will last long enough for
	httperf to sample.

2003-05-06  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* autobench:
	Now performs more thorough checking of options supplied in the
	config file or on the command line

2003-01-18  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* bench2graph:
	Fixed bug caused by hasty fix in rev 1.3 whereby the temporary
	output file was deleted immediately after creation.  Patch
	supplied by Steve Greenland <steve at lsli.com>.

2002-12-10  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* bench2graph: Output files are created in CWD once more.

2002-10-15  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* autobenchd:
	Fixed bug whereby httperf wouldn't be killed off when autobenchd
	receives "ABORT" and SIGTERMs its child.

2002-10-15  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* Released [autobench-2-0-1]

	* autobenchd:
	Fixed bug whereby httperf wouldn't be killed off when autobenchd
	receives "ABORT" and SIGTERMs its child.

2002-10-14  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* Released [autobench-2-0-0]

	* autobench_admin.1, autobenchd.1:
	Added man pages for autobench_admin and autobenchd.

	* bench2graph:
	Fixed bug whereby bench2graph would always move the output file to the
	CWD instead of putting it where you asked for it.

	* autobenchd:
	Tidied up debugging output, and added --verbose option for monitoring
	progress

	* autobenchd:
	Ensured that the child handles failed httperf runs correctly (by
	checking for 0 replies), and fixed get_command()'s handling of EAGAINS

	* autobench_admin:
	Admin controller for distributed autobench.  First fully
	functioning version.

	* autobench.conf:
	Added 'clients' config option for distributed autobench

2002-10-13  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* autobenchd:
	Daemon for distributed autobench.  Listens for connections from
	autobench_admin, and runs httperf against the target host, sending
	processed output back to autobench_admin for collation.

2002-10-12  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* debian/changelog, debian/conffiles, debian/control, debian/copyright, debian/dirs, debian/rules:
	Debianised ;-)

	* Makefile:
	Debianised - now set up either for building a Debian package or
	installing in /usr/local (or elsewhere) in the usual way.

	Added uninstall target which deletes everything.

	Now automatically updates autobench with the correct location of the
	master config file when PREFIX or ETCDIR are changed.

2002-05-02  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* Released [autobench-1-1-2]

	* autobench, autobench.1:
	Added a patch from Rob Fielding (rob at dsvr.net) which allows
	additional httperf options to be specified in the configuration file.

2001-07-16  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* bench2graph:
	In the interests of portability, changed #!/bin/bash to #!/bin/sh

2001-06-26  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* Released [autobench-1-1-1]

2001-06-26  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* autobench, autobench.1, autobench.conf:
	autobench now accepts --port1 and --port2 options to set the ports on
	which the servers should be contacted.  Both port1 and port2 are set
	to 80 by default in autobench.conf.

	Patches supplied by Daniel Serodio <daniel at ibnetwork.com.br>

2001-06-21  Julian T J Midgley  <jtjm@xenoclast.org>

	* Released [autobench-1-1-0]

2001-06-21  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* /home/jtjm/cvsroot/autobench/Makefile:
	Added benchgraph to install target

2001-06-20  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* /home/jtjm/cvsroot/autobench/bench2graph.1: Man page for bench2graph

	* /home/jtjm/cvsroot/autobench/bench2graph:
	Draws gnuplot graphs from autobench's output.
	Written by Samuel Dupas <samuel.dupas at synaptique.co.uk>

2001-05-11  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* /home/jtjm/cvsroot/autobench/Makefile, /home/jtjm/cvsroot/autobench/sesslog.1, /home/jtjm/cvsroot/autobench/sesslog.c:
	Added sesslog man page

2001-04-22  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* /home/jtjm/cvsroot/autobench/crfile.c:
	Simple fix to avoid warning during compile

2001-04-18  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* /home/jtjm/cvsroot/autobench/autobench:
	BUG FIX: --uri couldn't take an ampersand in its argument, and
	occasional division by zero errors were reported when httperf reported
	zero replies.  Both fixed.  Reported by Jocelyn
	Rakotomalala <jocelyn at 24pmteam.com>.

2001-04-17  Julian T. J. Midgley  <jtjm@xenoclast.org>

	* /home/jtjm/cvsroot/autobench/sesslog.c:
	Tidied up, and now defaults sensibly to use stdin and stdout when
	appropriate.


sesslog.c
/*******************************************************************
 * sesslog - creates a session log suitable for use with the       *
 * --wsesslog option of httperf from an NCSA Common Log Format     *
 * or combined log format log.                                     *
 *                                                                 *
 * Usage: sesslog [<infile>] [<outfile>]                           *
 *                                                                 *
 * Copyright (C) 2001 Julian T. J. Midgley <jtjm@xenoclast.org>    *
 *                                                                 *
 * This program is free software; you can redistribute it and/or   *
 * modify it under the terms of the GNU General Public License as  *
 * published by the Free Software Foundation; either version 2 of  *
 * the License, or (at your option) any later version.             *
 *                                                                 *
 * This program is distributed in the hope that it will be         *
 * useful, but WITHOUT ANY WARRANTY; without even the implied      * 
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR         *
 * PURPOSE.  See the GNU General Public License for more details.  *
 *                                                                 *
 * You should have received a copy of the GNU General Public       *
 * License along with this program; if not, write to the Free      *
 * Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,  *
 * MA  02111-1307  USA                                             *
 *                                                                 * 
 * A copy of version 2 of the GNU General Public License may be    * 
 * found in the file "LICENCE" in the autobench source tarball.    *
 *                                                                 *
 *******************************************************************/


#define BUFF_SIZE 1024
#include <stdio.h>

enum boolean { FALSE, TRUE }; 

int 
main(int argc, char *argv[])
{
  char *infile, *outfile;
  char method[BUFF_SIZE];
  char *p;
  int bm, em, bu, eu; /* Flags to indicate beginning and end of 
                             the HTTP method and the URL */
  FILE *ifs, *ofs;    /* Streams for input and output */
  char ch, pch;

  ifs = ofs = stderr;      /* Purely to avoid compile time errors of the  */
  infile = outfile = "";   /* "'foo' might be used uninitialized" variety */  

  switch (argc) {
  case 1: 
    ifs = stdin; ofs = stdout; break;
  case 2: 
    infile = argv[1]; ofs = stdout; break;
  case 3: 
    infile = argv[1]; outfile = argv[2]; break;
  default: 
    fprintf(stderr, "Usage: cr_sesslog [<infile>] [<outfile>]\n");
    return(-1); 
    break;
  }
  
  if (ifs != stdin && ((ifs = fopen(infile,"r")) == NULL)) {
    perror(infile);
    return(-1);
  }
  if (ofs != stdout && ((ofs = fopen(outfile,"w")) == NULL)) {
    perror(outfile);
    return(-1);
  }  
  
  bm = em = bu = eu = FALSE;
  pch = '*';
  p = method;

  /*****************************************************************************
   * The following is essentially the equivalent of the following Perl snippet *
   *     while (<IN>) {                                                        *
   *         if (m/\"(\S+) (\S+)/) {                                           *
   *             print OUT "$2 method=$1\n";                                   *
   *         }                                                                 *
   *     }                                                                     *
   * I've implemented this in C rather then Perl purely because it is          *
   * marginally faster.                                                        *
   *****************************************************************************/
  while ((ch = fgetc(ifs)) != EOF) {
    bm = ( bm || pch == '"');
    em = ( em || (bm && ch == ' '));
    bu = ( bu || (em && pch == ' '));
    eu = ( eu || (bu && ch == ' '));
    if ( bm && !em ) { 
      *p++ = ch;
    }
    if ( (bu && !eu) ) {
      fputc (ch, ofs);
    }
    pch = ch;
    if (ch == '\n') {
      *p = '\0';
      fputs(" method=",ofs);
      fputs (method, ofs);
      bm = em = bu = eu = FALSE;
      fputc(ch, ofs);
      p = method;
      pch = '*';
    }
  }
  
  fclose (ifs);
  fclose (ofs);
  return(0);
}


crfile.c
/*******************************************************************
 * crfile - create a file of the specified size containing random  *
 * ASCII characters.                                               *
 *                                                                 *
 * Copyright (C) 2001 Julian T. J. Midgley <jtjm@xenoclast.org>    *
 *                                                                 *
 * Usage: crfile -f <filename> -s <size> [-n]                      *
 *                                                                 *
 *  where <filename> is the name of the file to create             *
 *        <size>     is the size of the file in bytes              *
 *        -n         if present, specifies that the file should    *
 *                   be padded with nulls (enables the creation of *
 *                   large files that don't actually take up much  *
 *                   much space on disk)                           *
 *        -w[n]      Wrap the output at n characters, or at 80     *
 *                   characters if n omitted                       *
 *                                                                 *
 * This program is free software; you can redistribute it and/or   *
 * modify it under the terms of the GNU General Public License as  *
 * published by the Free Software Foundation; either version 2 of  *
 * the License, or (at your option) any later version.             *
 *                                                                 *
 * This program is distributed in the hope that it will be         *
 * useful, but WITHOUT ANY WARRANTY; without even the implied      * 
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR         *
 * PURPOSE.  See the GNU General Public License for more details.  *
 *                                                                 *
 * You should have received a copy of the GNU General Public       *
 * License along with this program; if not, write to the Free      *
 * Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,  *
 * MA  02111-1307  USA                                             *
 *                                                                 * 
 * A copy of version 2 of the GNU General Public License may be    * 
 * found in the file "LICENCE" in the autobench source tarball.    *
 *                                                                 *
 *******************************************************************/
          
#include <stdio.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdlib.h>
#include <time.h>

#define VERSION "1.0"
#define WORDWRAP 80
#define OPTSTRING "f:s:nw::"
#define BUFFER_SIZE 1048576

typedef struct {
  char buffer[BUFFER_SIZE]; 
  long ptr;                 /* Position of the next free space in the buffer */
  int flushed;              /* 1 if buffer empty, 0 otherwise                */
} wbuffer;

int
init_buffer(wbuffer *buf)
{
  /* Initialise the buffer, and set the flushed flag */
  buf->ptr = 0;
  buf->flushed = 1;
  return(1);
}


int
put_buffer(wbuffer *buf, char c)
{
  /* Push c onto the buffer, and reset the flushed flag */
  *(buf->buffer + buf->ptr) = c;
  buf->ptr++;
  buf->flushed = 0;
  return (buf->ptr == BUFFER_SIZE);
} 


int
write_buffer(wbuffer *buf, int fd)
{
  /* Write the contents of the buffer to file descriptor fd */
  write(fd, buf->buffer, buf->ptr);
  init_buffer(buf);
  return(1);
}


int
usage(void)
{
  const char * version = VERSION;
  fprintf(stderr,"crfile %s, by Julian T. J. Midgley <jtjm@xenoclast.org>\n\n",version);
  fprintf(stderr,"Usage: crfile -f <filename> -s <size> [-n -w[<l>]]\n");
  fprintf(stderr,"    -f <filename> : Create file <filename>\n");
  fprintf(stderr,"    -s <size>     : of size <size> bytes\n");
  fprintf(stderr,"    -n            : pad file with nulls\n");
  fprintf(stderr,"    -w[<l>]       : Wrap the output at <l> characters, or\n");
  fprintf(stderr,"                    80 characters if <l> omitted\n");
  return(0);
}
  

char
rand_ascii(void) 
{
  return(32+(int) ((rand()/(RAND_MAX+1.0))*95.0));
}

int
main(int argc, char **argv)
{
  char *filename; //, *optarg;
  long size = 0, i, l, line_length = WORDWRAP;
  char c;
  int fd, opt_f = 0, opt_s = 0, opt_n = 0, opt_w = 0;
  char opt = 0;
  wbuffer buf;

  filename = argv[0]; /* Initialised purely to avoid warnings on compile */

  /* Parse arguments */
  while (opt != -1) {
    opt = getopt(argc, argv, OPTSTRING);
    if (opt == 'f') {
      filename = optarg;
      opt_f = 1;
    }
    if (opt == 's') {
      size = strtol(optarg, NULL, 10);
      opt_s = 1;
    }
    if (opt == 'n') {
      opt_n = 1;
    }
    if (opt == 'w') {
      opt_w = 1;
      if (optarg) {
	line_length = strtol(optarg, NULL, 10);
	if (line_length < 2) {
	  fprintf(stderr, "Argument to -w must be greater than 2\n");
	  exit(1);
	}
      }
      else {
	line_length = WORDWRAP;
      }
    }
  }
  
  if (!(opt_s && opt_f)) {
    usage();
    exit(1);
  }
  
  printf("Creating file '%s', of size %ld bytes\n",filename,size);

  fd = open(filename, O_CREAT|O_WRONLY|O_EXCL,0666);
  if (fd == -1) {
    /* Error in opening file, fail gracefully */
    perror("Cannot open file");
    exit (1);
  }
  
  /* Initialise random number generator with current time */
  srand(time(NULL)+getpid()); 

  /* Initialise buffer */
  init_buffer(&buf);
  
  if (opt_n) {
    /* Create a file padded with nulls */
    c = rand_ascii();
    lseek(fd, size-1, SEEK_SET);
    write(fd, &c, 1);
  }
  else {
    /* Fill the file with random ASCII characters */
    l = 0;
    for (i=0; i<size; i++) {
      c = rand_ascii(); /* Generate ASCII character */
      if (put_buffer(&buf,c)) {
	write_buffer(&buf,fd);
      }
      l++;
      if (opt_w && (l > (line_length-2)) && (i < size-1)) {
	if (put_buffer(&buf,'\n')) {
	  write_buffer(&buf,fd);
	}
	l=0;
	i++;
      }
    }
  }
  
  /* If the buffer hasn't been flushed, flush it */
  if (! buf.flushed) {
    write_buffer(&buf,fd);
  }
  close (fd);  
  exit(0);
}
